crates/runtime/src/accelerated_table.rs

```diff
@@ -0,0 +1,260 @@
+    status,
+    timing::TimeMeasurement,
+};
+
+#[derive(Debug, Snafu)]
+enum Error {
+    #[snafu(display("Unable to get data from connector: {source}"))]
+    UnableToGetDataFromConnector { source: dataconnector::Error },
+
+    #[snafu(display("Unable to scan table provider: {source}"))]
+    UnableToScanTableProvider {
+        source: datafusion::error::DataFusionError,
+    },
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+// An accelerated table consists of a federated table and a local accelerator.
+//
+// The accelerator must support inserts.
+pub(crate) struct AcceleratedTable {
+    accelerator: Arc<dyn TableProvider>,
+    federated: Arc<dyn TableProvider>,
+    refresh_handle: JoinHandle<()>,
+}
+
+impl AcceleratedTable {
+    pub fn new(
+        dataset_name: String,
+        federated: Arc<dyn TableProvider>,
+        accelerator: Arc<dyn TableProvider>,
+        refresh_mode: RefreshMode,
+        refresh_interval: Option<Duration>,
+    ) -> Self {
+        let refresh_handle = tokio::spawn(Self::start_refresh(
+            dataset_name,
+            Arc::clone(&federated),
+            refresh_mode,
+            refresh_interval,
+            Arc::clone(&accelerator),
+        ));
+
+        Self {
+            accelerator,
+            federated,
+            refresh_handle,
+        }
+    }
+
+    async fn start_refresh(
+        dataset_name: String,
+        federated: Arc<dyn TableProvider>,
+        refresh_mode: RefreshMode,
+        refresh_interval: Option<Duration>,
+        accelerator: Arc<dyn TableProvider>,
+    ) {
+        let mut stream = Self::stream_updates(
+            dataset_name.clone(),
+            federated,
+            refresh_mode,
+            refresh_interval,
+        );
+        loop {
+            let future_result = stream.next().await;
+            match future_result {
+                Some(data_update) => {
+                    let data_update = match data_update {
+                        Ok(data_update) => data_update,
+                        Err(e) => {
+                            tracing::debug!("Error streaming data for {dataset_name}: {e}");
+                            break;
+                        }
+                    };
+                    let ctx = SessionContext::new();
+                    let state = ctx.state();
+
+                    let overwrite = data_update.update_type == UpdateType::Overwrite;
+                    match accelerator
+                        .insert_into(
+                            &state,
+                            Arc::new(DataUpdateExecutionPlan::new(data_update)),
+                            overwrite,
+                        )
+                        .await
+                    {
+                        Ok(plan) => {
+                            if let Err(e) = collect(plan, ctx.task_ctx()).await {
+                                tracing::error!("Error adding data for {dataset_name}: {e}");
```

Comment:
I think we would rather return the state of success/failure to the outside rather than swallow it in async code, and let the calling function decide what to do with the error.  one option is to make this an async function that returns an Error for example

```rust
 async fn start_refresh(
        dataset_name: String,
        federated: Arc<dyn TableProvider>,
        refresh_mode: RefreshMode,
        refresh_interval: Option<Duration>,
        accelerator: Arc<dyn TableProvider>,
    )  -> Option<TableRefreshError>
```

Reply:
Yeah, the design here is that we start this refresh on a "background thread" and immediately return the AcceleratedTable to be registered in DataFusion. In that design, there isn't anyone to return the error to - other than to report it to the console/metrics.

I'll keep it like this for this PR, since this has already taken way too much of my time and I want to get it merged. But happy to consider alternative ways to approach this.